diff --git a/src/utils/port.c b/src/utils/port.c
index c1228fd..b0c4221 100644
--- a/src/utils/port.c
+++ b/src/utils/port.c
@@ -355,7 +355,7 @@ void bctbx_set_self_thread_name(const char *name) {
 	prctl(PR_SET_NAME, name, NULL, NULL, NULL);
 #elif TARGET_OS_MAC
 	pthread_setname_np(name);
-#elif
+#else
 	bctbx_warning("bctbx_set_self_thread_name(): not implemented on this platform.");
 #endif
 }
@@ -1605,7 +1605,7 @@ void bctbx_sockaddr_remove_nat64_mapping(const struct sockaddr *v6, struct socka
 		if (htonl(0x0064ff9b) ==
 #ifdef _MSC_VER
 		    ((in6->sin6_addr.u.Word[0] << 16) & in6->sin6_addr.u.Word[1])
-#elif __APPLE__
+#elif defined(__APPLE__) || defined(__GENODE__)
 		    in6->sin6_addr.__u6_addr.__u6_addr32[0]
 #else
 		    in6->sin6_addr.s6_addr32[0]
