diff --git a/CMakeLists.txt b/CMakeLists.txt
index 779bef3..88624e4 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -74,6 +74,9 @@ set (CMAKE_CXX_STANDARD 17)
 
 set(CMAKE_CXX_FLAGS_BAK "${CMAKE_CXX_FLAGS}")
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")# need this for check_cxx_source_compiles because CMAKE_CXX_STANDARD doesn't work.
+if (GENODE)
+	set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wl,--whole-archive -Wl,-lgcc -Wl,--no-whole-archive")
+endif()
 check_cxx_source_compiles("#include <atomic>
 using namespace std;
 int main(int argc, char *argv[]) {
diff --git a/src/netsim.c b/src/netsim.c
index faa65bc..075ffde 100644
--- a/src/netsim.c
+++ b/src/netsim.c
@@ -78,7 +78,7 @@ static const char *sched_policy_to_string(int policy) {
 #endif
 
 static void set_high_prio(void) {
-#ifndef _WIN32
+#if !defined(_WIN32) && !defined(__GENODE__)
 	const char *sched_pref = getenv("ORTP_SIMULATOR_SCHED_POLICY");
 	int policy = SCHED_OTHER;
 	struct sched_param param;
