diff --git a/src/audiofilters/oss.c b/src/audiofilters/oss.c
index 81b38a6f..0335f69f 100644
--- a/src/audiofilters/oss.c
+++ b/src/audiofilters/oss.c
@@ -131,7 +131,11 @@ static int configure_fd(int fd, int bits, int stereo, int rate, int *minsz) {
 		res = ioctl(fd, SNDCTL_DSP_SETTRIGGER, &fl);
 		if (res < 0) ms_warning("OSS_TRIGGER: %s", strerror(errno));
 	}
+#ifdef __GENODE__
+	*minsz = 2048;
+#else
 	*minsz = min_size;
+#endif
 	return fd;
 }
 
@@ -253,8 +257,13 @@ static void oss_init(MSSndCard *card) {
 	d->read_started = FALSE;
 	d->write_started = FALSE;
 	d->bits = 16;
+#ifdef __GENODE__
+	d->rate = 44100;
+	d->stereo = TRUE;
+#else
 	d->rate = 8000;
 	d->stereo = FALSE;
+#endif
 	qinit(&d->rq);
 	d->bufferizer = ms_bufferizer_new();
 	ms_mutex_init(&d->mutex, NULL);
@@ -475,9 +484,11 @@ static void oss_write_process(MSFilter *f) {
 }
 
 static int set_rate(MSFilter *f, void *arg) {
+#ifndef __GENODE__
 	MSSndCard *card = (MSSndCard *)f->data;
 	OssData *d = (OssData *)card->data;
 	d->rate = *((int *)arg);
+#endif
 	return 0;
 }
 
@@ -489,15 +500,26 @@ static int get_rate(MSFilter *f, void *arg) {
 }
 
 static int set_nchannels(MSFilter *f, void *arg) {
+#ifndef __GENODE__
+	MSSndCard *card = (MSSndCard *)f->data;
+	OssData *d = (OssData *)card->data;
+	*((int *)arg) = d->rate;
+#endif
+	return 0;
+}
+
+
+static int get_nchannels(MSFilter *f, void *arg) {
 	MSSndCard *card = (MSSndCard *)f->data;
 	OssData *d = (OssData *)card->data;
-	d->stereo = (*((int *)arg) == 2);
+	*((int *)arg) = d->stereo ? 2 : 1;
 	return 0;
 }
 
 static MSFilterMethod oss_methods[] = {{MS_FILTER_SET_SAMPLE_RATE, set_rate},
                                        {MS_FILTER_GET_SAMPLE_RATE, get_rate},
                                        {MS_FILTER_SET_NCHANNELS, set_nchannels},
+                                       {MS_FILTER_GET_NCHANNELS, get_nchannels},
                                        {0, NULL}};
 
 MSFilterDesc oss_read_desc = {.id = MS_OSS_READ_ID,
