--- a/qemu/chardev/char-stdio.c
+++ b/qemu/chardev/char-stdio.c
@@ -104,8 +104,11 @@
     old_fd0_flags = fcntl(0, F_GETFL);
     tcgetattr(0, &oldtty);
     if (!g_unix_set_fd_nonblocking(0, true, NULL)) {
-        error_setg_errno(errp, errno, "Failed to set FD nonblocking");
+/*
+        error_setg_errno(errp, errno, "Failed to set FD nonblocking C");
         return;
+*/
+		printf("%s:%u ignore error nonblocking\n", __func__, __LINE__);
     }
     atexit(term_exit);
 
@@ -113,10 +116,13 @@
     act.sa_handler = term_stdio_handler;
     sigaction(SIGCONT, &act, NULL);
 
+	printf("%s:%u\n", __func__, __LINE__);
     qemu_chr_open_fd(chr, 0, 1);
 
     stdio_allow_signal = !opts->has_signal || opts->signal;
+	printf("%s:%u\n", __func__, __LINE__);
     qemu_chr_set_echo_stdio(chr, false);
+	printf("%s:%u\n", __func__, __LINE__);
 }
 #endif
 
