index ae7d0aa..c2a2eea 100644
--- a/meson.build
+++ b/meson.build
@@ -218,7 +218,7 @@ if host_cpu_family.startswith('x86')
         asm_sources = ['win64.S']
       endif
     endif
-  elif ['darwin', 'ios', 'linux', 'android'].contains(host_system)
+  elif ['darwin', 'ios', 'linux', 'android', 'freebsd'].contains(host_system)
     if size_t == 4
       if ['darwin', 'ios'].contains(host_system)
         TARGET = 'X86_DARWIN'
